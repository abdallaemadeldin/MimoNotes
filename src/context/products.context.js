import axios from "axios";
import React, { useState, createContext, useContext } from "react"

export const ProductsContext = createContext();

export const ProductsProvider = (props) => {
    const [loading, setLoading] = useState(true);
    const [categories, setCategories] = useState([]);
    const [selectedCategory, setSelectedCategory] = useState('All');
    const [products, setProducts] = useState([]);
    const [temp, setTemp] = useState([]);

    const toggleFav = (title) => {
        setProducts(st => [...st.map(e => e.title === title ? { ...e, isFav: !e.isFav } : e)]);
        setTemp(st => [...st.map(e => e.title === title ? { ...e, isFav: !e.isFav } : e)]);
    }

    const getCategories = () => axios.get('https://fakestoreapi.com/products/categories').then(res => setCategories(['All', ...res.data]));
    const getProducts = () => axios.get('https://fakestoreapi.com/products').then(res => {
        const items = [];
        res.data.map(e => { items.push({ ...e, isFav: false }); })
        setProducts(items);
        setTemp(items);
        setLoading(false);
    });

    const filter = (category) => {
        setSelectedCategory(category);
        if (category === 'All') setProducts(temp);
        else setProducts(temp.filter(product => product.category === category));
    }

    const contextValue = { loading, categories, selectedCategory, products, getCategories, getProducts, toggleFav, filter }

    return (
        <ProductsContext.Provider value={contextValue}>
            {props.children}
        </ProductsContext.Provider>
    );
}

export const useProductsContext = () => useContext(ProductsContext);