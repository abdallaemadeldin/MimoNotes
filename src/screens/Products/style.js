import { StyleSheet } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

export const style = () => {
    const { bottom } = useSafeAreaInsets();

    return StyleSheet.create({
        container: {
            flex: 1,
            backgroundColor: '#f1f1f1',
            alignItems: 'center'
        },
        tagView: {
            paddingHorizontal: 15,
            height: 30,
            backgroundColor: '#fff',
            marginHorizontal: 4,
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 15,
            marginVertical: 10
        },
        activeTag: {
            backgroundColor: 'orange'
        },
        tagLabel: {
            color: '#000',
            fontSize: 16,
        },
        activeTagLabel: {
            color: '#fff',
            fontSize: 16,
        },
        listHeader: {
            fontSize: 22,
            fontWeight: 'bold',
            color: '#000',
            marginBottom: 10,
            marginLeft: '2%'
        },
        productCard: {
            width: '46%',
            height: 300,
            marginHorizontal: '2%',
            marginBottom: 20
        },
        imgHolder: {
            width: '100%',
            height: '70%',
            backgroundColor: '#fff',
            borderRadius: 4,
            elevation: 2,
            shadowColor: '#000',
            shadowOffset: { width: 1, height: 1 },
            shadowOpacity: .5,
            shadowRadius: 4,
            padding: 10
        },
        productImg: {
            flex: 1,
        },
        contentContainerStyle: {
            paddingBottom: bottom + 10
        },
        productCardName: {
            color: '#000',
            marginTop: 30
        },
        productCardPrice: {
            color: '#000',
            fontSize: 16,
            fontWeight: 'bold',
            lineHeight: 26
        },
        floatingBtn: {
            width: 44,
            height: 44,
            borderRadius: 22,
            backgroundColor: '#fff',
            position: 'absolute',
            elevation: 5,
            shadowRadius: 5,
            shadowOpacity: .4,
            shadowOffset: { width: 1, height: 1 },
            shadowColor: '#000',
            bottom: 72,
            right: 10,
            justifyContent: 'center',
            alignItems: 'center'
        },
        indicator: {
            marginTop: '30%'
        }
    });
}