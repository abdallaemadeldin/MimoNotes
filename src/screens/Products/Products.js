import React, { memo } from 'react';
import { View, ScrollView, Image, FlatList, TouchableOpacity, Text, ActivityIndicator } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import MaskedView from '@react-native-masked-view/masked-view';
import { useProducts } from 'src/hooks';
import { Toolbar } from 'src/components/toolbar';
import { style } from './style';

const Products = () => {
    const { container, tagView, activeTag, tagLabel, activeTagLabel, indicator, listHeader, productCard, imgHolder, contentContainerStyle, productCardName, productCardPrice, productImg, floatingBtn } = style();
    const { categories, selectedCategory, products, loading, toggleFav, filter, ref, navigateTo } = useProducts();

    const renderItem = ({ item }) => {
        return (
            <TouchableOpacity style={productCard} activeOpacity={1} onPress={() => navigateTo("Details", { price: item.price, description: item.description })}>
                <View style={imgHolder}>
                    <Image source={{ uri: item.image }} style={productImg} resizeMode="center" />
                </View>
                <TouchableOpacity style={floatingBtn} activeOpacity={.8} onPress={() => toggleFav(item.title)}>
                    <MaskedView style={{
                        width: 24,
                        height: 24,
                    }} maskElement={<Icon name={item.isFav ? "heart" : "heart-outline"} size={24} />} key={Math.random()}>
                        <LinearGradient colors={['#21094E', '#511281', '#7952B3']} useAngle angle={45} style={{ flex: 1, height: '100%' }} />
                    </MaskedView>
                </TouchableOpacity>
                <View style={{
                    width: '100%',
                    height: '30%',
                }}>
                    <Text numberOfLines={2} style={productCardName}>{item.title}</Text>
                    <Text style={productCardPrice}>{`$${item.price}`}</Text>
                </View>
            </TouchableOpacity>
        );
    }

    return (
        <View style={container}>
            <Toolbar label="Clothing" />
            {loading ? <ActivityIndicator color="#21094E" style={indicator} /> : <>
                <ScrollView showsHorizontalScrollIndicator={false} horizontal>
                    {categories.map(category => {
                        const condition = category === selectedCategory;
                        return (
                            <TouchableOpacity key={category} style={[tagView, condition && activeTag]} onPress={() => {
                                filter(category);
                                ref.current.scrollToOffset({ animated: true, offset: 0 });
                            }}>
                                <Text style={condition ? activeTagLabel : tagLabel}>{category}</Text>
                            </TouchableOpacity>
                        );
                    })}
                </ScrollView>
                <FlatList
                    ref={ref}
                    data={products}
                    ListHeaderComponent={() => <Text style={listHeader}>{`${products.length} Items`}</Text>}
                    renderItem={renderItem}
                    numColumns={2}
                    removeClippedSubviews
                    windowSize={1}
                    getItemLayout={(_, index) => ({ length: 300, offset: 300 * index, index })}
                    keyExtractor={e => e.title}
                    showsVerticalScrollIndicator={false}
                    contentContainerStyle={contentContainerStyle}
                    style={{ width: '100%', height: '75%' }}
                />
            </>}
        </View>
    );
}

export default memo(Products);