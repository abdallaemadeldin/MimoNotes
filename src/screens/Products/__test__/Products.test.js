import React from 'react';
import renderer from 'react-test-renderer';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import Products from '../Products';
import { ProductsProvider } from 'src/context/products.context';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native");
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: jest.fn(),
            canGoBack: jest.fn(),
            goBack: jest.fn(),
        }),
    };
});

test('renders correctly', () => {
    const tree = renderer.create(
        <ProductsProvider>
            <SafeAreaProvider initialMetrics={{ insets: { top: 0, bottom: 0, left: 0, right: 0 } }}>
                <Products />
            </SafeAreaProvider>
        </ProductsProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});