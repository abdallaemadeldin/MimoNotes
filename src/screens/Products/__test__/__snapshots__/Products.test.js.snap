// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f1f1f1",
        "flex": 1,
      }
    }
  >
    <LinearGradient
      angle={45}
      colors={
        Array [
          "#21094E",
          "#511281",
          "#7952B3",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 44,
          "justifyContent": "center",
          "paddingTop": null,
          "width": 750,
        }
      }
      useAngle={true}
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Clothing
      </Text>
    </LinearGradient>
    <ActivityIndicator
      color="#21094E"
      style={
        Object {
          "marginTop": "30%",
        }
      }
    />
  </View>
</RNCSafeAreaProvider>
`;
